### PR TITLE
Make `i31ref` and `dataref` nullable

### PR DESCRIPTION
**Pull request by tlively**
Match the latest version of the GC spec. This change does not depend on V8
changing its interpretation of the shorthands because we are still temporarily
not emitting the binary shorthands, but all Binaryen users will have to update
their interpretations along with this change if they use the text or binary
shorthands.

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* main
  * **PR #4843** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/4843" target="_blank">Graphite</a>  👈
    * **PR #4844** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/4844" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/4843?utm_source=stack-comment).

**Comment by tlively** [comment]
@askeska-google has requested that we merge this after merging a solution for https://github.com/WebAssembly/binaryen/issues/4824.

File: CHANGELOG.md
Comment (kripken) [inline CHANGELOG.md:20]: Maybe add that this follows a spec change? (otherwise it seems like maybe it's some decision we made)

**Review by kripken — APPROVED** [review]

**Pull request merged**
